src/controllers/csi/driver/server.go

```diff
@@ -144,32 +146,38 @@ func (svr *CSIDriverServer) NodePublishVolume(ctx context.Context, req *csi.Node
 	} else if isMounted {
 		return &csi.NodePublishVolumeResponse{}, nil
 	}
-
-	volumeAttributes := req.GetVolumeContext()
-	if volumeAttributes[CSIVolumeAttributeName] == appvolumes.Mode {
-		log.Info("publishing app volume",
-			"target", volumeCfg.TargetPath,
-			"fstype", req.GetVolumeCapability().GetMount().GetFsType(),
-			"readonly", req.GetReadonly(),
-			"volumeID", volumeCfg.VolumeId,
-			"attributes", req.GetVolumeContext(),
-			"mountflags", req.GetVolumeCapability().GetMount().GetMountFlags(),
-		)
-		return svr.publishers[appvolumes.Mode].PublishVolume(ctx, volumeCfg)
+	publisher, ok := svr.publishers[volumeCfg.Mode]
+	if !ok {
+		return nil, status.Error(codes.Internal, fmt.Sprintf("unknown csi mode provided, mode=%s", volumeCfg.Mode))
 	}
-	return &csi.NodePublishVolumeResponse{}, nil
+	log.Info("publishing volume",
+		"csiMode", volumeCfg.Mode,
+		"target", volumeCfg.TargetPath,
+		"fstype", req.GetVolumeCapability().GetMount().GetFsType(),
+		"readonly", req.GetReadonly(),
+		"volumeID", volumeCfg.VolumeID,
+		"attributes", req.GetVolumeContext(),
+		"mountflags", req.GetVolumeCapability().GetMount().GetMountFlags(),
+	)
+	return publisher.PublishVolume(ctx, volumeCfg)
 }
 
 func (svr *CSIDriverServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpublishVolumeRequest) (*csi.NodeUnpublishVolumeResponse, error) {
 	volumeInfo, err := csivolumes.ParseNodeUnpublishVolumeRequest(req)
 	if err != nil {
 		return nil, err
 	}
-	response, err := svr.publishers[appvolumes.Mode].UnpublishVolume(ctx, volumeInfo)
-	if err != nil {
-		return nil, err
+	for _, publisher := range svr.publishers {
```

Comment:
is trying to unpublish the volume both ways the best idea here?
we should know how it was mounted by checking the database

Reply:
that is what actually happen, the unpublish checks if it has info about it in the database and if not it does nothing
(that is why https://github.com/Dynatrace/dynatrace-operator/pull/555#discussion_r805811276 happened)

Reply:
I think this shouldn't be hidden behind the `Unpublish` method
We should extract that check so it's more explicit, either as its own method on each publisher or as a check that's done before the actual unpublish

Reply:
>  either as its own method on each publisher

I like that